### ISSUE TITLE
Unable to resolve modules/files outside of root react-native directory in an ejected project / example monorepo request

### ISSUE THREAD

**Issue by dalinarkholin** [facebook/react-native#17391]
### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

Environment:
  OS: macOS Sierra 10.12.4
  Node: 8.5.0
  Yarn: 0.27.5
  npm: 5.3.0
  Watchman: 4.9.0
  Xcode: Xcode 8.2.1 Build version 8C1002
  Android Studio: 2.2 AI-145.3537739

Packages: (wanted => installed)
  react: 16.0.0 => 16.0.0
  react-native: 0.50.3 => 0.50.3

Target Platform: iOS

### Steps to Reproduce

(Write your steps here:)

Hard to write out reproduction steps since this request/bug might be outside of the scope of just react-native...but in any case, see the Readme of linked example repo to reproduce.

### Expected Behavior

I expected modules/files to resolve using aliases within a ejected React Native App, since they worked in a non-ejected(CRNA).

### Actual Behavior

My setup is pure typescript and includes api, native, core, and infrastructure code. It is setup using yarn workspaces and I am using `crna-make-symlinks-for-yarn-workspaces` to reference sym linked node_modules in root dir. It is possible this is the root issue, but wanted others input. I had it working with CRNA, but then when ejecting I got this module not found error again and because running `react-native run-ios` builds successfully I am fairly certain the issue is not related to the node_module linker but instead an aliasing issue.

Here is some more context.

<img width="292" alt="screen shot 2017-12-29 at 1 30 49 pm" src="https://user-images.githubusercontent.com/2343017/34444495-8eabe7d4-ec9c-11e7-97b6-d2506c22eb03.png">

package.json modules
```
"dependencies": {
    "react": "16.0.0",
    "react-native": "0.50.3"
  },
  "devDependencies": {
    "@types/react": "^16.0.28",
    "@types/react-native": "^0.51.0",
    "babel-plugin-module-resolver": "^3.0.0",
    "babel-preset-react-native-stage-0": "^1.0.1",
  }
```

babelrc (located in native) Note I need to create aliases to the compiled javascript from typescript.
```
{
  "presets": ["babel-preset-react-native-stage-0/decorator-support"],
  "plugins": ["transform-react-jsx-source",
    [
      "module-resolver",
      {
        "extensions": [".js", ".ios.js", ".android.js"],
        "alias": {
          "~/core": "../dist/core",
          "~/infrastructure": "../dist/infrastructure",
          "~/native": "../dist/native"
        },
        "cwd": "babelrc"
      }
    ]
  ]
}
```
What I am trying to import in my react native project

```
import test from "~/core";
alert(test);

compiles to:

const core_1 = require("~/core"); 
alert(core_1.default);
```

Error Im getting
<img width="354" alt="error" src="https://user-images.githubusercontent.com/2343017/34445192-35d77a48-eca0-11e7-89e8-248bbc17a662.png">

This is a fraction of what I have tried:
- Adding `{ name: "topLevelPath" }` to the top level package.json. It works (example `import core from "topLevelPath/dist/core"`, but I can't then alias it.
- Removing and adding root in babelrc. Didn't work.
- Adding and removing `"cwd": "babelrc"`. Didn't work.
- Tried using module-alias instead of babel-plugin-module-resolver. Didn't work.
- Adding `react-native` as the preset. Didn't work.
- I reset `yarn start -- --reset-cache` before every build.
- Tried deleting the ios build folder.

My goal is a pure typescript monorepo. Hopefully someone sees something here I have missed.

### Reproducible Demo

https://github.com/dalinarkholin/example-typescript-monorepo

**Comment by dalinarkholin** [comment]
To answer my own question, during the eject process for react-native the babelrc preset gets changed from `babel-preset-expo` to `babel-preset-react-native-stage-0/decorator-support`

Simple reverting this back fixed the issue. Not sure if any harm will come of it by leaving it as `babel-preset-expo`

Any thoughts from the react native folks? Thanks for your time.

**Comment by dalinarkholin** [comment]
I created a custom rn-cli-config.js and got it working. If anyone else runs into this issue you can see what I did here:

https://github.com/dalinarkholin/example-typescript-monorepo

**Comment by fishmwei** [comment]
Is it resolved?  I meet it in version 0.54.2

**Comment by ashconnell** [comment]
Future note to myself and others...

yarn workspaces now support the nohoist option so you can set up your root packages like this

```
{
  "private": true,
  "workspaces": {
    "packages": ["packages/*"],
    "nohoist": ["**/react-native", "**/react-native/**"]
  }
}
```

here's the blog post: https://yarnpkg.com/blog/2018/02/15/nohoist/
